test/distributed/org/apache/cassandra/distributed/test/sai/PartialUpdateHandlingTest.java

```diff
@@ -0,0 +1,401 @@
+ * particular, cases where writes do not propagate to all replicas can, if not resolved correctly, lead to consistency
+ * violations, as rows that should not appear in our results do, while rows that should appear do not.
+ * <p> 
+ * The variables that affect the behavior of these queries are, at minimum, the following:
+ * <p> 
+ * 1.) The combination of write & read consistency levels used by the client. Reads at ONE/LOCAL_ONE trivially
+ *     avoid having to resolve data from diverging replicas.
+ * 2.) Interaction of existing data (if there is any), with partial updates and deletes. The same query that
+ *     erroneously returns stale matches with naïve resolution and partial updates on top of existing data might 
+ *     fail to return live matches with partial updates and no previous data.
+ * 3.) The number of query clauses and their targets. Clauses may target partition keys, clustering keys, static 
+ *     columns, and regular columns, and some combinations are more problematic than others.
+ * 4.) The repaired state of SSTables that participate in the query. A fully repaired on-disk set of SSTables cannot
+ *     produce erroneous results due to split rows (i.e. rows which contain partial updates of different columns
+ *     across different partitions).
+ * 5.) Whether data resides in SSTables or Memtables. The latter is implicitly unrepaired.
+ * 6.) Interaction w/ existing mechanisms on the distributed read path that deal with short reads, replica filtering
+ *     protection, etc.
+ * 7.) The relationship between columns selected and columns restricted by queries. (If coordinator filtering is
+ *     involved at the implementation level, retrieving enough information to do that filtering is important.)
+ * 8.) The timestamps of partial updates and deletes, especially for single-column queries that might produce
+ *     stale matches if not resolved correctly.
+ */
+@RunWith(Parameterized.class)
+public class PartialUpdateHandlingTest extends TestBaseImpl
+{
+    private static final String TEST_TABLE_NAME = "test_partial_updates";
+
+    private static Cluster CLUSTER;
+
+    @BeforeClass
+    public static void setUpCluster() throws IOException
+    {
+        CLUSTER = init(Cluster.build(3).withConfig(config -> config.set("hinted_handoff_enabled", false).with(GOSSIP).with(NETWORK)).start());
+
+        // All parameterized test scenarios share the same table and attached indexes, but write to different partitions
+        // that are deleted after each scenario completes.
+        String createTableDDL = String.format("CREATE TABLE %s.%s (pk int, ck int, s int static, a int, b int, PRIMARY KEY (pk, ck)) WITH read_repair = 'NONE'",
+                                              KEYSPACE, TEST_TABLE_NAME);
+        CLUSTER.schemaChange(createTableDDL);
+        CLUSTER.disableAutoCompaction(KEYSPACE);
+
+        CLUSTER.schemaChange(String.format("CREATE INDEX ck_idx ON %s.%s(ck) USING 'sai'", KEYSPACE, TEST_TABLE_NAME));
```

Comment:
I think we are missing indexes on composite partition key columns

Reply:
Yeah, I don't think it would be too hard to add this.